byzcoin/struct.go

```diff
@@ -1,12 +1,531 @@
+	}
+
+	return b.Bucket(sid)
+}
+
+// setMaxSize enables the cleaning of old state changes when the storage
+// size is above a given threshold. Note that the value is not strict.
+func (s *stateChangeStorage) setMaxSize(size int) {
+	s.maxSize = size
+}
+
+// setMaxNbrBlock enables the cleaning of state changes belonging
+// to blocks with an old index.
+func (s *stateChangeStorage) setMaxNbrBlock(nbr int) {
+	s.maxNbrBlock = nbr
+}
+
+// This will clean the oldest state changes when the total size
+// is above the maximum. It will remove elements until cleanThreshold of
+// the space is available.
+func (s *stateChangeStorage) cleanBySize() error {
+	if s.size < s.maxSize || s.maxSize == 0 {
+		// nothing to clean
+		return nil
+	}
+
+	err := s.db.Update(func(tx *bolt.Tx) error {
+		// We make enough space to not have to do it everytime
+		// we append state changes
+		thres := int(float64(s.maxSize) * cleanThreshold)
+
+		for s.size > thres {
+			if len(s.sortedKeys) == 0 {
+				// This should never happen
+				return errors.New("Nothing can be cleaned, something is wrong with the storage implementation")
+			}
+
+			// Get the oldest version with the same instance ID
+			iid := s.sortedKeys[0].key[:prefixLength]
+
+			b := s.getBucket(tx, s.sortedKeys[0].scid)
+			c := b.Cursor()
+
+			k, v := c.Seek(iid)
+			if !bytes.HasPrefix(k, iid) {
+				// This should never occur
+				return errors.New("Missing key in the storage")
+			}
+
+			err := c.Delete()
+			if err != nil {
+				return err
+			}
+
+			s.size -= len(v)
+			k, v = c.Next()
+
+			if bytes.HasPrefix(k, iid) {
+				var sce StateChangeEntry
+				err := protobuf.Decode(v, &sce)
+				if err != nil {
+					return err
+				}
+				s.sortedKeys[0].time = sce.Timestamp
+				copy(s.sortedKeys[0].key, k)
+
+				sort.Sort(s.sortedKeys)
+			} else {
+				// if none, that means it was the last
+				s.sortedKeys = s.sortedKeys[1:]
+			}
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+// This will clean the state changes per instance where the block
+// index is too low compared to the threshold
+func (s *stateChangeStorage) cleanByBlock(scs StateChanges, sb *skipchain.SkipBlock) error {
+	if s.maxNbrBlock == 0 {
+		return nil
+	}
+
+	thres := int64(sb.Index - s.maxNbrBlock)
+
+	return s.db.Update(func(tx *bolt.Tx) error {
+		b := s.getBucket(tx, sb.SkipChainID())
+
+		// Prevent from cleaning the same instance twice
+		done := map[string]bool{}
+
+		// Clean only the instances where state changes have been added
+		for _, sc := range scs {
+			_, ok := done[string(sc.InstanceID)]
+			if !ok {
+				done[string(sc.InstanceID)] = true
+
+				var buf bytes.Buffer
+				err := binary.Write(&buf, binary.BigEndian, thres)
+				if err != nil {
+					return err
+				}
+				index := buf.Bytes()
+
+				c := b.Cursor()
+				for k, v := c.Seek(sc.InstanceID); k != nil && bytes.HasPrefix(k, sc.InstanceID); k, v = c.Next() {
+					if bytes.Compare(k[len(k)-len(index):], index) <= 0 {
+						c.Delete()
```

Comment:
will this return an error? we should check it if it does

Reply:
Yes it will, I need to fix that